File: backend/internal/repository/acl.go
package repository

import (
	"context"
	"time"

	model "github.com/docvault/backend/internal/domain/document"
)

// VisibilityParams identifies a single document and the principal whose read
// visibility is being evaluated.
type VisibilityParams struct {
	TenantID   string
	OrgID      string
	DocumentID string
	UserID     string
	GroupIDs   []string
	IsAdmin    bool
}

// FolderVisibilityParams identifies a single folder and the principal whose read
// visibility is being evaluated.
type FolderVisibilityParams struct {
	TenantID string
	OrgID    string
	FolderID string
	UserID   string
	GroupIDs []string
	IsAdmin  bool
}

// ListVisibleParams filters a per-principal visible document listing.
type ListVisibleParams struct {
	TenantID string
	OrgID    string
	UserID   string
	GroupIDs []string
	IsAdmin  bool
	DocType  string
	FolderID string
	Status   string
	Language string
	Cursor   string
	Limit    int
}

// Group is an org-scoped principal grouping.
type Group struct {
	ID        string
	TenantID  string
	OrgID     string
	Name      string
	CreatedBy *string
	CreatedAt time.Time
}

// Grant is a per-resource read/write/delete permission for a principal.
type Grant struct {
	ID            string
	TenantID      string
	OrgID         string
	ResourceType  string
	ResourceID    string
	PrincipalType string
	PrincipalID   string
	Permission    string
	GrantedBy     *string
	CreatedAt     time.Time
}

// CreateGroupParams creates an org-scoped group.
type CreateGroupParams struct {
	TenantID  string
	OrgID     string
	Name      string
	CreatedBy *string
}

// CreateGrantParams creates a per-resource grant for a principal.
type CreateGrantParams struct {
	TenantID      string
	OrgID         string
	ResourceType  string
	ResourceID    string
	PrincipalType string
	PrincipalID   string
	Permission    string
	GrantedBy     *string
}

// ResourceRef identifies a single ACL-managed resource within an org.
type ResourceRef struct {
	TenantID     string
	OrgID        string
	ResourceType string
	ResourceID   string
}

// SetRestrictedParams toggles the is_restricted flag on a document or folder.
type SetRestrictedParams struct {
	TenantID   string
	OrgID      string
	ResourceID string
	Restricted bool
}

// The ACL surface is segregated into role-specific ports so consumers can
// depend on just the slice they use (interface segregation). The concrete
// adapter satisfies every port, and ACLRepository composes them for the
// composition root and the few callers that genuinely span roles.

// DocVisibilityPort evaluates per-document read/write visibility for a
// principal. The retrieval/authorization seam depends on just this.
type DocVisibilityPort interface {
	IsDocumentVisible(ctx context.Context, params VisibilityParams) (bool, error)
	IsDocumentWritable(ctx context.Context, params VisibilityParams) (bool, error)
}

// FolderVisibilityPort evaluates per-folder read visibility for a principal.
type FolderVisibilityPort interface {
	IsFolderVisible(ctx context.Context, params FolderVisibilityParams) (bool, error)
}

// GroupMembershipPort resolves the groups a user belongs to within an org. It
// is read on nearly every request that evaluates visibility.
type GroupMembershipPort interface {
	ListUserGroupIDs(ctx context.Context, userID, orgID string) ([]string, error)
}

// DocumentListingPort lists the documents a principal may read.
type DocumentListingPort interface {
	ListVisibleDocuments(ctx context.Context, params ListVisibleParams) ([]model.Document, *string, error)
}

// GroupAdminPort manages org-scoped groups and their membership.
type GroupAdminPort interface {
	CreateGroup(ctx context.Context, params CreateGroupParams) (Group, error)
	DeleteGroup(ctx context.Context, tenantID, orgID, groupID string) (int64, error)
	ListGroups(ctx context.Context, tenantID, orgID string) ([]Group, error)
	AddGroupMember(ctx context.Context, tenantID, orgID, groupID, userID string) error
	RemoveGroupMember(ctx context.Context, tenantID, orgID, groupID, userID string) (int64, error)
}

// GrantPort manages per-resource permission grants.
type GrantPort interface {
	CreateGrant(ctx context.Context, params CreateGrantParams) (string, error)
	GrantTargetExists(ctx context.Context, ref ResourceRef) (bool, error)
	DeleteGrant(ctx context.Context, tenantID, orgID, grantID string) (int64, error)
	ListGrantsByResource(ctx context.Context, ref ResourceRef) ([]Grant, error)
	DeleteGrantsForResource(ctx context.Context, ref ResourceRef) (int64, error)
}

// RestrictionPort toggles the is_restricted flag on documents and folders.
type RestrictionPort interface {
	SetDocumentRestricted(ctx context.Context, params SetRestrictedParams) (int64, error)
	SetFolderRestricted(ctx context.Context, params SetRestrictedParams) (int64, error)
}

// ACLRepository composes every ACL role port. It is the producer-side superset
// satisfied by the concrete postgres adapter; prefer depending on the narrow
// role ports above at each consumer.
type ACLRepository interface {
	DocVisibilityPort
	FolderVisibilityPort
	GroupMembershipPort
	DocumentListingPort
	GroupAdminPort
	GrantPort
	RestrictionPort
}

File: web/lib/api/acl.ts
import { apiFetch } from './core';

export type AclResourceType = 'document' | 'folder';
export type AclPrincipalType = 'user' | 'group';
export type AclPermission = 'read' | 'write' | 'delete';

// The backend serializes repository.Grant without json tags, so the list
// endpoint returns capitalized Go field names (ID/PrincipalType/...). Match
// that shape exactly, as lib/api/casbin.ts does for MemberPermission.
export interface Grant {
  ID: string;
  TenantID: string;
  OrgID: string;
  ResourceType: AclResourceType;
  ResourceID: string;
  PrincipalType: AclPrincipalType;
  PrincipalID: string;
  Permission: AclPermission;
  GrantedBy: string | null;
  CreatedAt: string;
}

export interface GrantListResponse {
  grants: Grant[];
}

export interface CreateGrantInput {
  resource_type: AclResourceType;
  resource_id: string;
  principal_type: AclPrincipalType;
  principal_id: string;
  permission?: AclPermission;
}

export interface CreateGrantResponse {
  id: string;
  resource_type: AclResourceType;
  resource_id: string;
  principal_type: AclPrincipalType;
  principal_id: string;
  permission: AclPermission;
}

export interface RestrictResponse {
  id: string;
  restricted: boolean;
}

// The backend serializes repository.Group without json tags, so groups come
// back with capitalized Go field names (ID/Name/...). Match that shape.
export interface Group {
  ID: string;
  TenantID: string;
  OrgID: string;
  Name: string;
  CreatedBy: string | null;
  CreatedAt: string;
}

export interface GroupListResponse {
  groups: Group[];
}

// --- Restrictions ---

export async function restrictDocument(id: string): Promise<RestrictResponse> {
  return apiFetch<RestrictResponse>(`/documents/${id}/restrict`, {
    method: 'PUT',
  });
}

export async function unrestrictDocument(
  id: string
): Promise<RestrictResponse> {
  return apiFetch<RestrictResponse>(`/documents/${id}/restrict`, {
    method: 'DELETE',
  });
}

export async function restrictFolder(id: string): Promise<RestrictResponse> {
  return apiFetch<RestrictResponse>(`/folders/${id}/restrict`, {
    method: 'PUT',
  });
}

export async function unrestrictFolder(id: string): Promise<RestrictResponse> {
  return apiFetch<RestrictResponse>(`/folders/${id}/restrict`, {
    method: 'DELETE',
  });
}

// --- Grants ---

export async function listGrants(
  resourceType: AclResourceType,
  resourceId: string
): Promise<GrantListResponse> {
  const params = new URLSearchParams({
    resource_type: resourceType,
    resource_id: resourceId,
  });
  const response = await apiFetch<GrantListResponse>(
    `/acl/grants?${params.toString()}`
  );
  return {
    ...response,
    grants: response?.grants ?? [],
  };
}

export async function createGrant(
  input: CreateGrantInput
): Promise<CreateGrantResponse> {
  return apiFetch<CreateGrantResponse>('/acl/grants', {
    method: 'POST',
    body: JSON.stringify(input),
  });
}

export async function deleteGrant(id: string): Promise<void> {
  await apiFetch(`/acl/grants/${id}`, { method: 'DELETE' });
}

// --- Groups ---

export async function listGroups(): Promise<GroupListResponse> {
  const response = await apiFetch<GroupListResponse>('/groups');
  return {
    ...response,
    groups: response?.groups ?? [],
  };
}

export async function createGroup(name: string): Promise<{ group: Group }> {
  return apiFetch<{ group: Group }>('/groups', {
    method: 'POST',
    body: JSON.stringify({ name }),
  });
}

export async function deleteGroup(id: string): Promise<void> {
  await apiFetch(`/groups/${id}`, { method: 'DELETE' });
}

export async function addGroupMember(
  groupId: string,
  userId: string
): Promise<void> {
  await apiFetch(`/groups/${groupId}/members`, {
    method: 'POST',
    body: JSON.stringify({ user_id: userId }),
  });
}

export async function removeGroupMember(
  groupId: string,
  userId: string
): Promise<void> {
  await apiFetch(`/groups/${groupId}/members/${userId}`, {
    method: 'DELETE',
  });
}
